# 2.7 Experiments

## (a)

In [1]:
from cs336_basics.tokenizer import Tokenizer
import time

In [2]:
def pretty_output(corpus: str, token_ids: list[int], preview_len: int, elapsed_time: float | None =None):
    corpus_bytes = corpus.encode()

    print("[Tokenizer] encode complete")
    print(f"    -Corpus length (bytes): {len(corpus_bytes)}")
    print(f"    -Token count: {len(token_ids)}")
    print(f"    -Compression ratio: {len(corpus_bytes) / len(token_ids)}")

    if elapsed_time:
        print(f"    -Elapsed time: {elapsed_time:.4f} seconds")
        print(f"    -Throughput: {len(corpus_bytes) / elapsed_time} bytes/second")

    print(f"    - Preview (corpus): {corpus[:preview_len]}...")

    print(f"    - Preview (token ids): {token_ids}")

In [3]:
vocab_path = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/vocab_table/tiny_story/output_TinyStoriesV2-GPT4-train_serialization_vocab_20251010_112414.json"
merge_path = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/vocab_table/tiny_story/output_TinyStoriesV2-GPT4-train_serialization_merge_20251010_112414.json"

corpus_path = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/sample_data/sample_TinyStoriesV2-GPT4-train_k10.txt"

# ---- Start timing ----

start_time = time.perf_counter()

tokenizer = Tokenizer.from_files(vocab_path, merge_path, ["<|endoftext|>"])

with open(corpus_path) as f_corpus:
    corpus = f_corpus.read()

token_ids = tokenizer.encode(corpus)

# ---- End timing ----
end_time = time.perf_counter()
elapsed_time = end_time - start_time

# print(f"[Tokenizer] encodeing complete in {elapsed_time:.4f} seconds")
pretty_output(corpus, token_ids, 30, elapsed_time)


[Tokenizer] encode complete
    -Corpus length (bytes): 7736
    -Token count: 1794
    -Compression ratio: 4.312151616499443
    -Elapsed time: 24.6217 seconds
    -Throughput: 314.19496468679097 bytes/second
    - Preview (corpus): 
Once upon a time, there was a...
    - Preview (token ids): [10, 430, 439, 259, 398, 44, 401, 283, 259, 346, 44, 779, 497, 46, 424, 364, 661, 2760, 267, 1301, 46, 527, 327, 44, 259, 390, 418, 550, 266, 1421, 354, 263, 497, 46, 285, 418, 652, 266, 2753, 354, 259, 1685, 44, 408, 309, 283, 876, 46, 285, 418, 1098, 1441, 267, 452, 365, 920, 367, 46, 10, 919, 263, 418, 283, 5809, 44, 309, 534, 420, 3724, 354, 828, 2463, 46, 285, 3724, 283, 365, 731, 44, 267, 263, 390, 418, 407, 309, 266, 480, 573, 46, 712, 44, 263, 418, 1137, 828, 2295, 266, 4765, 263, 3724, 46, 285, 3724, 427, 573, 44, 267, 263, 418, 283, 376, 46, 10, 1273, 263, 3724, 283, 1368, 44, 263, 418, 1098, 1441, 266, 2753, 354, 263, 1685, 46, 285, 779, 497, 283, 378, 797, 267, 452, 365, 2432, 46, 172

**Answer:** the TinyStories tokenizer’s compression ratio is 4.312151616499443

## (b)

In [4]:
vocab_path_b = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/vocab_table/tiny_story/output_TinyStoriesV2-GPT4-train_serialization_vocab_20251010_112414.json"
merge_path_b = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/vocab_table/tiny_story/output_TinyStoriesV2-GPT4-train_serialization_merge_20251010_112414.json"

corpus_path = "/Users/lucas/Documents/GitHub/stf-assignment1-basics/cs336_basics/smoke_test_fixture/sample_data/sample_owt_valid_k10.txt"

tokenizer = Tokenizer.from_files(vocab_path_b, merge_path_b, ["<|endoftext|>"])

with open(corpus_path) as f_corpus:
    corpus = f_corpus.read()

token_ids = tokenizer.encode(corpus)

# print(f"Encode complete: corpus length(in bytes)={len(corpus_bytes)}, corpus preview: {corpus[:30]}, token ids length={len(token_ids)}, token ids={token_ids}")

pretty_output(corpus, token_ids, 30)

[Tokenizer] encode complete
    -Corpus length (bytes): 73517
    -Token count: 21555
    -Compression ratio: 3.4106703781025285
    - Preview (corpus): In the second game of last yea...
    - Preview (token ids): [1463, 263, 4656, 1026, 370, 2212, 1572, 384, 732, 793, 5917, 776, 291, 1281, 5169, 298, 961, 615, 270, 2895, 44, 2771, 2611, 865, 299, 944, 67, 550, 713, 992, 2159, 7465, 45, 48, 354, 557, 103, 9436, 832, 267, 550, 484, 370, 328, 1355, 346, 494, 638, 653, 437, 313, 670, 284, 270, 463, 944, 270, 110, 1102, 384, 883, 370, 346, 494, 638, 653, 7154, 44, 266, 2771, 286, 8166, 543, 6759, 586, 3098, 352, 109, 869, 44, 2611, 865, 299, 1459, 263, 308, 275, 351, 291, 670, 121, 3662, 111, 266, 259, 7156, 2457, 579, 1033, 325, 46, 10, 10, 1100, 44, 267, 362, 4038, 263, 670, 121, 3662, 111, 1406, 770, 7465, 57, 1341, 99, 377, 370, 9131, 851, 354, 263, 327, 46, 1354, 384, 1687, 558, 46, 10, 10, 2896, 384, 576, 308, 275, 351, 291, 384, 3069, 8790, 313, 263, 8273, 6108, 503, 516, 58, 10, 10

**Answer:** Compression ratio: 3.4106703781025285

## (c)

In [6]:
Pile_in_gb = 825
Pile_in_bytes = Pile_in_gb * 1024 ** 3

estimated_encoding_time4Pile_in_sec = Pile_in_bytes / 314.19496468679097
estimated_encoding_time4Pile_in_days = estimated_encoding_time4Pile_in_sec / (60 * 60 * 24)
estimated_encoding_time4Pile_in_year = estimated_encoding_time4Pile_in_days / 365

print(f"Estimated time required for the Pile dataset (825GB of text): {estimated_encoding_time4Pile_in_sec} seconds, equivalent to {estimated_encoding_time4Pile_in_days} days, equivalent to {estimated_encoding_time4Pile_in_year} years")

Estimated time required for the Pile dataset (825GB of text): 2819386382.2198973 seconds, equivalent to 32631.78683124881 days, equivalent to 89.40215570205153 years


**Answer:** Throughput: 310.6875414523819 bytes/second;  It would take 2819386382.2198973 seconds to tokenize the Pile dataset (825GB of text), equivalent to 32631.78683124881 days, equivalent to 89.40215570205153 years